gpuallocator/allocator.go

```diff
@@ -45,12 +45,14 @@ func NewBestEffortAllocator() (*Allocator, error) {
 
 // NewAllocator creates a new Allocator using the given allocation policy
 func NewAllocator(policy Policy) (*Allocator, error) {
-	err := nvml.Init()
-	if err != nil {
-		return nil, fmt.Errorf("error initializing NVML: %v", err)
+	nvmllib := nvml.New()
+	if ret := nvmllib.Init(); ret != nvml.SUCCESS {
+		return nil, fmt.Errorf("error initializing NVML: %v", ret)
 	}
 
-	devices, err := NewDevices()
+	devices, err := NewDevices(
+		WithNvmlLib(nvmllib),
+	)
 	if err != nil {
 		return nil, fmt.Errorf("error enumerating GPU devices: %v", err)
```

Comment:
Since NVML was initialized prior to this, should we invoke `nvmllib.Shutdown()` here?

Reply:
Note that we add the `nvmllib.Shutdown()` call to the runtime finalizer at the end of this fuction. I suppose in the cases where `NewDevices` returns an error this will not be triggered and we may want to add it first. I didn't want to modify this logic as part of this PR though.

Reply:
Fair enough.